File: solution/143ReorderList/143ReorderList.cpp
#include <sstream>
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <unordered_map>
#include <algorithm>
#include <limits>

using namespace std;

/*
https://oj.leetcode.com/problems/reorder-list/

Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You must do this in-place without altering the nodes' values.

For example,
Given {1,2,3,4}, reorder it to {1,4,2,3}.
*/

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    void reorderList(ListNode *head) {
    	int size = 0;
    	ListNode *p=head;
    	while(p!=nullptr){
    		p=p->next;
    		size++;
    	}
    	if(size<=2) return;
    	
    	int mid = size/2;
        int i = 0;
        ListNode *m = head;
        p = nullptr;
        while(m!=nullptr&&i<mid){
        	p = m;
        	m = m->next;
        	i++;
        }
        ListNode *q = m;
        ListNode *r = q->next;
        while(r!=nullptr){
        	p = q;
        	q = r;
        	r = r->next;
        	q->next = p;
        }
        m->next = nullptr;
        p = head;
        ListNode *pn = p->next;
        ListNode *qn = q->next;

        while(p!=m&&q!=m){
		    p->next = q;
		    q->next = pn;
		    p = pn;
		    q = qn;
		    pn= pn->next;
		    qn= qn->next;
		}
    }
};

int main(){
	return 0;
}

File: src/143ReorderList/143ReorderList.go
package main

import "fmt"

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func reorderList(head *ListNode) {
	if head == nil {
		return
	}
	n := 0
	p := head
	for p != nil {
		p = p.Next
		n++
	}
	m := (0 + n - 1) / 2
	mid := head
	for m != 0 {
		mid = mid.Next
		m--
	}

	var q, r *ListNode
	p = mid
	q = p.Next
	for q != nil {
		r = q.Next
		q.Next = p
		p = q
		q = r
	}
	mid.Next = nil

	q = head
	for q != nil {
		r = q.Next
		o := p.Next
		q.Next = p
		p.Next = r
		q = r
		p = o
	}
	return
}

func main() {
	next := &ListNode{2, nil}
	head := &ListNode{1, next}
	reorderList(head)
	for head != nil {
		fmt.Printf("%d ", head.Val)
		head = head.Next
	}
}
